fuzzer/src/main.rs

```diff
@@ -0,0 +1,443 @@
+//! This is a fuzzer for pulldown-cmark, which tries to detect super-linear parsing behaviour.
+//! Additionally it can detect panicks of pulldown-cmark.
+//!
+//! The general idea is to extract relevant literals from the pulldown-cmark source code
+//! (see the `literals` module).
+//! Short random combinations of literals are generated (like concatenating 6 random literals),
+//! which are called patterns.
+//! Those patterns are repeated until they reach a byte-length of `NUM_BYTES`.
+//! We iterate over subslices of the pattern to get `SAMPLE_SIZE` uniformly distributed
+//! (len, time)-pairs.
+//! Those pairs are fed into a scoring function to test if they seem to be linear or might be
+//! super-linear (see the `scoring` module).
+//! All of this is performed in parallel on multiple threads.
+//!
+//! Unfortunately we don't live in a perfect world, where linear patterns produce a perfect line.
+//! instead we need to deal with large outliers, e.g. if a thread is rescheduled while performing
+//! time measurements.
+//! Therefore, we have a sort-of two pass system.
+//! The first pass doesn't perform any outlier detection / handling.
+//! This is because outlier handling is pretty slow.
+//! The first pass just assumes there aren't any outliers, calculates `SAMPLE_SIZE` points and
+//! applies the scoring functions to them.
+//! This filters out most patterns already.
+//! If however possible non-linear behaviour is found, we perform the second pass.
+//! In the second pass, we double the `SAMPLE_SIZE`, because more points make for higher precision
+//! in the scoring functions.
+//! Additionally, we assume that the (len, time)-pairs interpreted as coordinates on a graph  are
+//! strictly monotonously increasing.
+//! If we encounter a time-value, which is larger than the next time value, that'll probably be an
+//! outlier.
+//! As such, if we encounter such a value, we'll re-time it.
+//! Unfortunately, again, we have an edge-case.
+//! When testing substrings of the input, we might not always end up after the whole pattern,
+//! but instead in the middle of it.
```

Comment:
I guess we *might* avoid that problem this way, but at the same time I think that it is valuable to find these kinds of weird parsing behaviour like #287 and #312. If you think that they add no / not enough value, it'd also be possible to change it to align to pattern-length instead of character-length.